lib/bike_brigade_web/components/sms_message_list_component.ex

```diff
@@ -0,0 +1,139 @@
+defmodule BikeBrigadeWeb.Components.SMSMessageListComponent do
+  use Phoenix.Component
+  use BikeBrigadeWeb, :html
+  alias BikeBrigade.Messaging.SmsMessage
+
+
+  def preview_message(%SmsMessage{body: body}) when not is_nil(body) do
+    if String.length(body) > 30 do
+      String.slice(body, 0, 30) <> "..."
+    else
+      body
+    end
+  end
+
+  def preview_message(%SmsMessage{media: media}) when length(media) > 0 do
+    "Attachment: #{length(media)} media"
+  end
+
+  def preview_message(_), do: "unknown message"
+
+
+  def handle_build_message_link({"/messages", rider_id}) do
+    ~p"/messages/#{rider_id}"
+  end
+
+  def handle_build_message_link({"/campaign", campaign_id, rider_id}) do
+    ~p"/campaigns/#{campaign_id}/messaging/riders/#{rider_id}"
+  end
+
+
+  # TODO: can :any be a tuple?
+  # TODO: fix indent.
+  attr :sms_list, :any, required: true
+
+  @doc """
+  A helper component to determine where clicking on a rider in an sms list takes us
+  we render the sms list in different capacities; this allows deciding to which sms
+  instance clicking a rider would take us to.
+  """
+  def message_link(assigns) do
+    route = case assigns.sms_list do
+      {"/messages"} -> ~p"/messages/#{assigns.rider_id}"
+      {"/campaigns", campaign_id} -> ~p"/campaigns/#{campaign_id}/messaging/riders/#{assigns.rider_id}"
+    end
```

Comment:
This can be better.

We're passing `sms_list` to the component, which seems to only exist to help us determine the url we go to when we click the rider. The name `sms_list` doesn't really point to that, and the shape of the data is unclear - a tuple of a string or a string and number.

I think a better solution is to pass a `rider_link_fn` to the component which would take a `rider_id`, so the caller would make a message list component as either
```
 <.sms_message_list
      rider_link_fn={fn (rider_id) -> ~p"/messages/#{rider_id}" end}
      conversations={@conversations}
```

Or
```
<.sms_message_list
      rider_link_fn={fn (rider_id) -> ~p"/campaigns/#{campaign_id}/messaging/riders/#{rider_id}" end}
      conversations={@conversations}
      
```

Note in the second example the caller knows the campaign_id

Reply:
I thought it wasn't possible to pass lambda's to components, but it turns out I was incorrectly invoking them (forgot you need a `.` -> `rider_link_fn.()`.

Now it works 🎈 !